tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/RawCustomGradient.java

```diff
@@ -0,0 +1,56 @@
+/*
+ Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow.op;
+
+import java.util.List;
+import org.tensorflow.GraphOperation;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.TensorFlow;
+import org.tensorflow.internal.c_api.GradFunc;
+
+/**
+ * A custom gradient for an op of unspecified type. Should be registered using {@link
+ * TensorFlow#registerCustomGradient(String, RawCustomGradient)}.
+ *
+ * <p>The op type of {@code op} will depend on the op type string passed to the registration method.
+ * Note that the registration method can be called more than once, resulting this gradient function
+ * being used for multiple different op types.
+ */
+@FunctionalInterface
+public interface RawCustomGradient {
```

Comment:
Why doesn't this have a typing relationship with `CustomGradient`?

Reply:
Because `CustomGradient` takes the op input class, and `RawCustomGradient` takes the `GraphOperation` class, and there's no relationship between them.

I suppose I could have something like `ICustoomGradient<T>` but I don't really see the point.

Reply:
Why is the one that operates on `RawOpInputs` called `CustomGradient` and the one which operates on `GraphOperation` called `RawCustomGradient`? Maybe `RawOpInputs` should be `OpInputs`? I'm not clear why it's raw, beyond it being an abstract class.

Reading the two methods in `org.tensorflow.TensorFlow` it looks like `CustomGradient` is only for ops with classes, and `RawCustomGradient` allows you to make one for arbitrary graph operations, however the docs for the two classes don't state this. If that's the correct interpretation then it should be in the docs for the interfaces, and if it's not then whatever they are for should be in the docs.

Reply:
You can use either for ops with classes, it's primarily there as a backup or if you can't use the reflection based Op class wrapping for some reason.  You can also re-use `RawCustomGradient` for multiple ops.

I named it `Raw` as a contrast to the typed `CustomGradient`, similar to `RawOp`.  I'm not attached to the name if you have a better idea though.